core/block_validator.go

```diff
@@ -0,0 +1,228 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/pow"
+	"gopkg.in/fatih/set.v0"
+)
+
+type BlockValidator struct {
+	// Canonical block chain
+	bc *BlockChain
+	// Proof of work used for validating
+	Pow pow.PoW
+}
+
+// NewBlockValidator returns a new block validator which is safe for re-use
+func NewBlockValidator(blockchain *BlockChain, pow pow.PoW) *BlockValidator {
+	sm := &BlockValidator{
+		Pow: pow,
+		bc:  blockchain,
+	}
+	return sm
+}
+
+func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+	if v.bc.HasBlock(block.Hash()) {
+		return &KnownBlockError{block.Number(), block.Hash()}
+	}
+	if !v.bc.HasBlock(block.ParentHash()) {
+		return ParentError(block.ParentHash())
+	}
+	var (
+		parent = v.bc.GetBlock(block.ParentHash())
+		header = block.Header()
+	)
+	// validate the block header
+	if err := ValidateHeader(v.Pow, header, parent.Header(), false, false); err != nil {
+		return err
+	}
+	// varify the uncles are correctly rewarded
+	if err := v.VerifyUncles(block, parent); err != nil {
+		return err
+	}
+
+	// Verify UncleHash before running other uncle validations
+	unclesSha := types.CalcUncleHash(block.Uncles())
+	if unclesSha != header.UncleHash {
+		return fmt.Errorf("invalid uncles root hash. received=%x calculated=%x", header.UncleHash, unclesSha)
+	}
+
+	// The transactions Trie's root (R = (Tr [[i, RLP(T1)], [i, RLP(T2)], ... [n, RLP(Tn)]]))
+	// can be used by light clients to make sure they've received the correct Txs
+	txSha := types.DeriveSha(block.Transactions())
+	if txSha != header.TxHash {
+		return fmt.Errorf("invalid transaction root hash. received=%x calculated=%x", header.TxHash, txSha)
+	}
+
+	return nil
+}
+
+func (v *BlockValidator) ValidateState(block, parent *types.Block, state *state.StateDB, receipts types.Receipts, usedGas *big.Int) (batch ethdb.Batch, err error) {
+	var (
+		header = block.Header()
+	)
+	if block.GasUsed().Cmp(usedGas) != 0 {
+		err = ValidationError(fmt.Sprintf("gas used error (%v / %v)", block.GasUsed(), usedGas))
```

Comment:
This may be a style question, but I'm a bit against naked return values. Sometimes they can indeed be helpful when the return paths can be complicated with various combinations of the output values being set/unset (or if the return values are inspected by a deferred recovery function). Here however the method's return is either `nil, someError` or `result, nil` (with no return inspection), so IMHO it would be much more readable to use classical returns (i.e. `return nil, fmt,Errorf(...)` throughout, and drop the return variable names).